// mediasoup/node/lib/fbs/consumer/consumer-layers.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsumerLayersT = exports.ConsumerLayers = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
class ConsumerLayers {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsConsumerLayers(bb, obj) {
        return (obj || new ConsumerLayers()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsConsumerLayers(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new ConsumerLayers()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    spatialLayer() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : 0;
    }
    temporalLayer() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : null;
    }
    static startConsumerLayers(builder) {
        builder.startObject(2);
    }
    static addSpatialLayer(builder, spatialLayer) {
        builder.addFieldInt8(0, spatialLayer, 0);
    }
    static addTemporalLayer(builder, temporalLayer) {
        builder.addFieldInt8(1, temporalLayer, null);
    }
    static endConsumerLayers(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createConsumerLayers(builder, spatialLayer, temporalLayer) {
        ConsumerLayers.startConsumerLayers(builder);
        ConsumerLayers.addSpatialLayer(builder, spatialLayer);
        if (temporalLayer !== null)
            ConsumerLayers.addTemporalLayer(builder, temporalLayer);
        return ConsumerLayers.endConsumerLayers(builder);
    }
    unpack() {
        return new ConsumerLayersT(this.spatialLayer(), this.temporalLayer());
    }
    unpackTo(_o) {
        _o.spatialLayer = this.spatialLayer();
        _o.temporalLayer = this.temporalLayer();
    }
}
exports.ConsumerLayers = ConsumerLayers;
class ConsumerLayersT {
    spatialLayer;
    temporalLayer;
    constructor(spatialLayer = 0, temporalLayer = null) {
        this.spatialLayer = spatialLayer;
        this.temporalLayer = temporalLayer;
    }
    pack(builder) {
        return ConsumerLayers.createConsumerLayers(builder, this.spatialLayer, this.temporalLayer);
    }
}
exports.ConsumerLayersT = ConsumerLayersT;
