// mediasoup/node/lib/fbs/consumer/layers-change-notification.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LayersChangeNotificationT = exports.LayersChangeNotification = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
const consumer_layers_1 = require("../../fbs/consumer/consumer-layers");
class LayersChangeNotification {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsLayersChangeNotification(bb, obj) {
        return (obj || new LayersChangeNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsLayersChangeNotification(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new LayersChangeNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    layers(obj) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? (obj || new consumer_layers_1.ConsumerLayers()).__init(this.bb.__indirect(this.bb_pos + offset), this.bb) : null;
    }
    static startLayersChangeNotification(builder) {
        builder.startObject(1);
    }
    static addLayers(builder, layersOffset) {
        builder.addFieldOffset(0, layersOffset, 0);
    }
    static endLayersChangeNotification(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createLayersChangeNotification(builder, layersOffset) {
        LayersChangeNotification.startLayersChangeNotification(builder);
        LayersChangeNotification.addLayers(builder, layersOffset);
        return LayersChangeNotification.endLayersChangeNotification(builder);
    }
    unpack() {
        return new LayersChangeNotificationT((this.layers() !== null ? this.layers().unpack() : null));
    }
    unpackTo(_o) {
        _o.layers = (this.layers() !== null ? this.layers().unpack() : null);
    }
}
exports.LayersChangeNotification = LayersChangeNotification;
class LayersChangeNotificationT {
    layers;
    constructor(layers = null) {
        this.layers = layers;
    }
    pack(builder) {
        const layers = (this.layers !== null ? this.layers.pack(builder) : 0);
        return LayersChangeNotification.createLayersChangeNotification(builder, layers);
    }
}
exports.LayersChangeNotificationT = LayersChangeNotificationT;
