// mediasoup/node/lib/fbs/consumer/score-notification.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScoreNotificationT = exports.ScoreNotification = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
const consumer_score_1 = require("../../fbs/consumer/consumer-score");
class ScoreNotification {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsScoreNotification(bb, obj) {
        return (obj || new ScoreNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsScoreNotification(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new ScoreNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    score(obj) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? (obj || new consumer_score_1.ConsumerScore()).__init(this.bb.__indirect(this.bb_pos + offset), this.bb) : null;
    }
    static startScoreNotification(builder) {
        builder.startObject(1);
    }
    static addScore(builder, scoreOffset) {
        builder.addFieldOffset(0, scoreOffset, 0);
    }
    static endScoreNotification(builder) {
        const offset = builder.endObject();
        builder.requiredField(offset, 4); // score
        return offset;
    }
    static createScoreNotification(builder, scoreOffset) {
        ScoreNotification.startScoreNotification(builder);
        ScoreNotification.addScore(builder, scoreOffset);
        return ScoreNotification.endScoreNotification(builder);
    }
    unpack() {
        return new ScoreNotificationT((this.score() !== null ? this.score().unpack() : null));
    }
    unpackTo(_o) {
        _o.score = (this.score() !== null ? this.score().unpack() : null);
    }
}
exports.ScoreNotification = ScoreNotification;
class ScoreNotificationT {
    score;
    constructor(score = null) {
        this.score = score;
    }
    pack(builder) {
        const score = (this.score !== null ? this.score.pack(builder) : 0);
        return ScoreNotification.createScoreNotification(builder, score);
    }
}
exports.ScoreNotificationT = ScoreNotificationT;
