// mediasoup/node/lib/fbs/sctp-parameters/sctp-parameters.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SctpParametersT = exports.SctpParameters = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
class SctpParameters {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsSctpParameters(bb, obj) {
        return (obj || new SctpParameters()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsSctpParameters(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new SctpParameters()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    port() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint16(this.bb_pos + offset) : 0;
    }
    maxSendMessageSize() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    maxReceiveMessageSize() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    sendBufferSize() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    perStreamSendQueueLimit() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    maxReceiverWindowBufferSize() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    isDataChannel() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    static startSctpParameters(builder) {
        builder.startObject(7);
    }
    static addPort(builder, port) {
        builder.addFieldInt16(0, port, 0);
    }
    static addMaxSendMessageSize(builder, maxSendMessageSize) {
        builder.addFieldInt32(1, maxSendMessageSize, 0);
    }
    static addMaxReceiveMessageSize(builder, maxReceiveMessageSize) {
        builder.addFieldInt32(2, maxReceiveMessageSize, 0);
    }
    static addSendBufferSize(builder, sendBufferSize) {
        builder.addFieldInt32(3, sendBufferSize, 0);
    }
    static addPerStreamSendQueueLimit(builder, perStreamSendQueueLimit) {
        builder.addFieldInt32(4, perStreamSendQueueLimit, 0);
    }
    static addMaxReceiverWindowBufferSize(builder, maxReceiverWindowBufferSize) {
        builder.addFieldInt32(5, maxReceiverWindowBufferSize, 0);
    }
    static addIsDataChannel(builder, isDataChannel) {
        builder.addFieldInt8(6, +isDataChannel, +false);
    }
    static endSctpParameters(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createSctpParameters(builder, port, maxSendMessageSize, maxReceiveMessageSize, sendBufferSize, perStreamSendQueueLimit, maxReceiverWindowBufferSize, isDataChannel) {
        SctpParameters.startSctpParameters(builder);
        SctpParameters.addPort(builder, port);
        SctpParameters.addMaxSendMessageSize(builder, maxSendMessageSize);
        SctpParameters.addMaxReceiveMessageSize(builder, maxReceiveMessageSize);
        SctpParameters.addSendBufferSize(builder, sendBufferSize);
        SctpParameters.addPerStreamSendQueueLimit(builder, perStreamSendQueueLimit);
        SctpParameters.addMaxReceiverWindowBufferSize(builder, maxReceiverWindowBufferSize);
        SctpParameters.addIsDataChannel(builder, isDataChannel);
        return SctpParameters.endSctpParameters(builder);
    }
    unpack() {
        return new SctpParametersT(this.port(), this.maxSendMessageSize(), this.maxReceiveMessageSize(), this.sendBufferSize(), this.perStreamSendQueueLimit(), this.maxReceiverWindowBufferSize(), this.isDataChannel());
    }
    unpackTo(_o) {
        _o.port = this.port();
        _o.maxSendMessageSize = this.maxSendMessageSize();
        _o.maxReceiveMessageSize = this.maxReceiveMessageSize();
        _o.sendBufferSize = this.sendBufferSize();
        _o.perStreamSendQueueLimit = this.perStreamSendQueueLimit();
        _o.maxReceiverWindowBufferSize = this.maxReceiverWindowBufferSize();
        _o.isDataChannel = this.isDataChannel();
    }
}
exports.SctpParameters = SctpParameters;
class SctpParametersT {
    port;
    maxSendMessageSize;
    maxReceiveMessageSize;
    sendBufferSize;
    perStreamSendQueueLimit;
    maxReceiverWindowBufferSize;
    isDataChannel;
    constructor(port = 0, maxSendMessageSize = 0, maxReceiveMessageSize = 0, sendBufferSize = 0, perStreamSendQueueLimit = 0, maxReceiverWindowBufferSize = 0, isDataChannel = false) {
        this.port = port;
        this.maxSendMessageSize = maxSendMessageSize;
        this.maxReceiveMessageSize = maxReceiveMessageSize;
        this.sendBufferSize = sendBufferSize;
        this.perStreamSendQueueLimit = perStreamSendQueueLimit;
        this.maxReceiverWindowBufferSize = maxReceiverWindowBufferSize;
        this.isDataChannel = isDataChannel;
    }
    pack(builder) {
        return SctpParameters.createSctpParameters(builder, this.port, this.maxSendMessageSize, this.maxReceiveMessageSize, this.sendBufferSize, this.perStreamSendQueueLimit, this.maxReceiverWindowBufferSize, this.isDataChannel);
    }
}
exports.SctpParametersT = SctpParametersT;
