// mediasoup/node/lib/fbs/web-rtc-transport/dtls-state-change-notification.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DtlsStateChangeNotificationT = exports.DtlsStateChangeNotification = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
const dtls_state_1 = require("../../fbs/web-rtc-transport/dtls-state");
class DtlsStateChangeNotification {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsDtlsStateChangeNotification(bb, obj) {
        return (obj || new DtlsStateChangeNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsDtlsStateChangeNotification(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new DtlsStateChangeNotification()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    dtlsState() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.readUint8(this.bb_pos + offset) : dtls_state_1.DtlsState.NEW;
    }
    remoteCert(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    static startDtlsStateChangeNotification(builder) {
        builder.startObject(2);
    }
    static addDtlsState(builder, dtlsState) {
        builder.addFieldInt8(0, dtlsState, dtls_state_1.DtlsState.NEW);
    }
    static addRemoteCert(builder, remoteCertOffset) {
        builder.addFieldOffset(1, remoteCertOffset, 0);
    }
    static endDtlsStateChangeNotification(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createDtlsStateChangeNotification(builder, dtlsState, remoteCertOffset) {
        DtlsStateChangeNotification.startDtlsStateChangeNotification(builder);
        DtlsStateChangeNotification.addDtlsState(builder, dtlsState);
        DtlsStateChangeNotification.addRemoteCert(builder, remoteCertOffset);
        return DtlsStateChangeNotification.endDtlsStateChangeNotification(builder);
    }
    unpack() {
        return new DtlsStateChangeNotificationT(this.dtlsState(), this.remoteCert());
    }
    unpackTo(_o) {
        _o.dtlsState = this.dtlsState();
        _o.remoteCert = this.remoteCert();
    }
}
exports.DtlsStateChangeNotification = DtlsStateChangeNotification;
class DtlsStateChangeNotificationT {
    dtlsState;
    remoteCert;
    constructor(dtlsState = dtls_state_1.DtlsState.NEW, remoteCert = null) {
        this.dtlsState = dtlsState;
        this.remoteCert = remoteCert;
    }
    pack(builder) {
        const remoteCert = (this.remoteCert !== null ? builder.createString(this.remoteCert) : 0);
        return DtlsStateChangeNotification.createDtlsStateChangeNotification(builder, this.dtlsState, remoteCert);
    }
}
exports.DtlsStateChangeNotificationT = DtlsStateChangeNotificationT;
