// mediasoup/node/lib/fbs/transport/close-data-producer-request.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloseDataProducerRequestT = exports.CloseDataProducerRequest = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
class CloseDataProducerRequest {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsCloseDataProducerRequest(bb, obj) {
        return (obj || new CloseDataProducerRequest()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsCloseDataProducerRequest(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new CloseDataProducerRequest()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    dataProducerId(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    static startCloseDataProducerRequest(builder) {
        builder.startObject(1);
    }
    static addDataProducerId(builder, dataProducerIdOffset) {
        builder.addFieldOffset(0, dataProducerIdOffset, 0);
    }
    static endCloseDataProducerRequest(builder) {
        const offset = builder.endObject();
        builder.requiredField(offset, 4); // data_producer_id
        return offset;
    }
    static createCloseDataProducerRequest(builder, dataProducerIdOffset) {
        CloseDataProducerRequest.startCloseDataProducerRequest(builder);
        CloseDataProducerRequest.addDataProducerId(builder, dataProducerIdOffset);
        return CloseDataProducerRequest.endCloseDataProducerRequest(builder);
    }
    unpack() {
        return new CloseDataProducerRequestT(this.dataProducerId());
    }
    unpackTo(_o) {
        _o.dataProducerId = this.dataProducerId();
    }
}
exports.CloseDataProducerRequest = CloseDataProducerRequest;
class CloseDataProducerRequestT {
    dataProducerId;
    constructor(dataProducerId = null) {
        this.dataProducerId = dataProducerId;
    }
    pack(builder) {
        const dataProducerId = (this.dataProducerId !== null ? builder.createString(this.dataProducerId) : 0);
        return CloseDataProducerRequest.createCloseDataProducerRequest(builder, dataProducerId);
    }
}
exports.CloseDataProducerRequestT = CloseDataProducerRequestT;
