// mediasoup/node/lib/fbs/transport/options.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.OptionsT = exports.Options = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
class Options {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsOptions(bb, obj) {
        return (obj || new Options()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsOptions(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new Options()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    direct() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    initialAvailableOutgoingBitrate() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : null;
    }
    enableSctp() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    maxSendMessageSize() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    maxReceiveMessageSize() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    sctpSendBufferSize() {
        const offset = this.bb.__offset(this.bb_pos, 14);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    sctpPerStreamSendQueueLimit() {
        const offset = this.bb.__offset(this.bb_pos, 16);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    sctpMaxReceiverWindowBufferSize() {
        const offset = this.bb.__offset(this.bb_pos, 18);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    isDataChannel() {
        const offset = this.bb.__offset(this.bb_pos, 20);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    static startOptions(builder) {
        builder.startObject(9);
    }
    static addDirect(builder, direct) {
        builder.addFieldInt8(0, +direct, +false);
    }
    static addInitialAvailableOutgoingBitrate(builder, initialAvailableOutgoingBitrate) {
        builder.addFieldInt32(1, initialAvailableOutgoingBitrate, null);
    }
    static addEnableSctp(builder, enableSctp) {
        builder.addFieldInt8(2, +enableSctp, +false);
    }
    static addMaxSendMessageSize(builder, maxSendMessageSize) {
        builder.addFieldInt32(3, maxSendMessageSize, 0);
    }
    static addMaxReceiveMessageSize(builder, maxReceiveMessageSize) {
        builder.addFieldInt32(4, maxReceiveMessageSize, 0);
    }
    static addSctpSendBufferSize(builder, sctpSendBufferSize) {
        builder.addFieldInt32(5, sctpSendBufferSize, 0);
    }
    static addSctpPerStreamSendQueueLimit(builder, sctpPerStreamSendQueueLimit) {
        builder.addFieldInt32(6, sctpPerStreamSendQueueLimit, 0);
    }
    static addSctpMaxReceiverWindowBufferSize(builder, sctpMaxReceiverWindowBufferSize) {
        builder.addFieldInt32(7, sctpMaxReceiverWindowBufferSize, 0);
    }
    static addIsDataChannel(builder, isDataChannel) {
        builder.addFieldInt8(8, +isDataChannel, +false);
    }
    static endOptions(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createOptions(builder, direct, initialAvailableOutgoingBitrate, enableSctp, maxSendMessageSize, maxReceiveMessageSize, sctpSendBufferSize, sctpPerStreamSendQueueLimit, sctpMaxReceiverWindowBufferSize, isDataChannel) {
        Options.startOptions(builder);
        Options.addDirect(builder, direct);
        if (initialAvailableOutgoingBitrate !== null)
            Options.addInitialAvailableOutgoingBitrate(builder, initialAvailableOutgoingBitrate);
        Options.addEnableSctp(builder, enableSctp);
        Options.addMaxSendMessageSize(builder, maxSendMessageSize);
        Options.addMaxReceiveMessageSize(builder, maxReceiveMessageSize);
        Options.addSctpSendBufferSize(builder, sctpSendBufferSize);
        Options.addSctpPerStreamSendQueueLimit(builder, sctpPerStreamSendQueueLimit);
        Options.addSctpMaxReceiverWindowBufferSize(builder, sctpMaxReceiverWindowBufferSize);
        Options.addIsDataChannel(builder, isDataChannel);
        return Options.endOptions(builder);
    }
    unpack() {
        return new OptionsT(this.direct(), this.initialAvailableOutgoingBitrate(), this.enableSctp(), this.maxSendMessageSize(), this.maxReceiveMessageSize(), this.sctpSendBufferSize(), this.sctpPerStreamSendQueueLimit(), this.sctpMaxReceiverWindowBufferSize(), this.isDataChannel());
    }
    unpackTo(_o) {
        _o.direct = this.direct();
        _o.initialAvailableOutgoingBitrate = this.initialAvailableOutgoingBitrate();
        _o.enableSctp = this.enableSctp();
        _o.maxSendMessageSize = this.maxSendMessageSize();
        _o.maxReceiveMessageSize = this.maxReceiveMessageSize();
        _o.sctpSendBufferSize = this.sctpSendBufferSize();
        _o.sctpPerStreamSendQueueLimit = this.sctpPerStreamSendQueueLimit();
        _o.sctpMaxReceiverWindowBufferSize = this.sctpMaxReceiverWindowBufferSize();
        _o.isDataChannel = this.isDataChannel();
    }
}
exports.Options = Options;
class OptionsT {
    direct;
    initialAvailableOutgoingBitrate;
    enableSctp;
    maxSendMessageSize;
    maxReceiveMessageSize;
    sctpSendBufferSize;
    sctpPerStreamSendQueueLimit;
    sctpMaxReceiverWindowBufferSize;
    isDataChannel;
    constructor(direct = false, initialAvailableOutgoingBitrate = null, enableSctp = false, maxSendMessageSize = 0, maxReceiveMessageSize = 0, sctpSendBufferSize = 0, sctpPerStreamSendQueueLimit = 0, sctpMaxReceiverWindowBufferSize = 0, isDataChannel = false) {
        this.direct = direct;
        this.initialAvailableOutgoingBitrate = initialAvailableOutgoingBitrate;
        this.enableSctp = enableSctp;
        this.maxSendMessageSize = maxSendMessageSize;
        this.maxReceiveMessageSize = maxReceiveMessageSize;
        this.sctpSendBufferSize = sctpSendBufferSize;
        this.sctpPerStreamSendQueueLimit = sctpPerStreamSendQueueLimit;
        this.sctpMaxReceiverWindowBufferSize = sctpMaxReceiverWindowBufferSize;
        this.isDataChannel = isDataChannel;
    }
    pack(builder) {
        return Options.createOptions(builder, this.direct, this.initialAvailableOutgoingBitrate, this.enableSctp, this.maxSendMessageSize, this.maxReceiveMessageSize, this.sctpSendBufferSize, this.sctpPerStreamSendQueueLimit, this.sctpMaxReceiverWindowBufferSize, this.isDataChannel);
    }
}
exports.OptionsT = OptionsT;
