// mediasoup/node/lib/fbs/transport/socket-flags.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketFlagsT = exports.SocketFlags = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
class SocketFlags {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsSocketFlags(bb, obj) {
        return (obj || new SocketFlags()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsSocketFlags(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new SocketFlags()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    ipv6Only() {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    udpReusePort() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? !!this.bb.readInt8(this.bb_pos + offset) : false;
    }
    static startSocketFlags(builder) {
        builder.startObject(2);
    }
    static addIpv6Only(builder, ipv6Only) {
        builder.addFieldInt8(0, +ipv6Only, +false);
    }
    static addUdpReusePort(builder, udpReusePort) {
        builder.addFieldInt8(1, +udpReusePort, +false);
    }
    static endSocketFlags(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createSocketFlags(builder, ipv6Only, udpReusePort) {
        SocketFlags.startSocketFlags(builder);
        SocketFlags.addIpv6Only(builder, ipv6Only);
        SocketFlags.addUdpReusePort(builder, udpReusePort);
        return SocketFlags.endSocketFlags(builder);
    }
    unpack() {
        return new SocketFlagsT(this.ipv6Only(), this.udpReusePort());
    }
    unpackTo(_o) {
        _o.ipv6Only = this.ipv6Only();
        _o.udpReusePort = this.udpReusePort();
    }
}
exports.SocketFlags = SocketFlags;
class SocketFlagsT {
    ipv6Only;
    udpReusePort;
    constructor(ipv6Only = false, udpReusePort = false) {
        this.ipv6Only = ipv6Only;
        this.udpReusePort = udpReusePort;
    }
    pack(builder) {
        return SocketFlags.createSocketFlags(builder, this.ipv6Only, this.udpReusePort);
    }
}
exports.SocketFlagsT = SocketFlagsT;
