// mediasoup/node/lib/fbs/rtp-parameters/encoding-mapping.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EncodingMappingT = exports.EncodingMapping = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
class EncodingMapping {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsEncodingMapping(bb, obj) {
        return (obj || new EncodingMapping()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsEncodingMapping(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new EncodingMapping()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    rid(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    ssrc() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : null;
    }
    scalabilityMode(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    mappedSsrc() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    static startEncodingMapping(builder) {
        builder.startObject(4);
    }
    static addRid(builder, ridOffset) {
        builder.addFieldOffset(0, ridOffset, 0);
    }
    static addSsrc(builder, ssrc) {
        builder.addFieldInt32(1, ssrc, null);
    }
    static addScalabilityMode(builder, scalabilityModeOffset) {
        builder.addFieldOffset(2, scalabilityModeOffset, 0);
    }
    static addMappedSsrc(builder, mappedSsrc) {
        builder.addFieldInt32(3, mappedSsrc, 0);
    }
    static endEncodingMapping(builder) {
        const offset = builder.endObject();
        return offset;
    }
    static createEncodingMapping(builder, ridOffset, ssrc, scalabilityModeOffset, mappedSsrc) {
        EncodingMapping.startEncodingMapping(builder);
        EncodingMapping.addRid(builder, ridOffset);
        if (ssrc !== null)
            EncodingMapping.addSsrc(builder, ssrc);
        EncodingMapping.addScalabilityMode(builder, scalabilityModeOffset);
        EncodingMapping.addMappedSsrc(builder, mappedSsrc);
        return EncodingMapping.endEncodingMapping(builder);
    }
    unpack() {
        return new EncodingMappingT(this.rid(), this.ssrc(), this.scalabilityMode(), this.mappedSsrc());
    }
    unpackTo(_o) {
        _o.rid = this.rid();
        _o.ssrc = this.ssrc();
        _o.scalabilityMode = this.scalabilityMode();
        _o.mappedSsrc = this.mappedSsrc();
    }
}
exports.EncodingMapping = EncodingMapping;
class EncodingMappingT {
    rid;
    ssrc;
    scalabilityMode;
    mappedSsrc;
    constructor(rid = null, ssrc = null, scalabilityMode = null, mappedSsrc = 0) {
        this.rid = rid;
        this.ssrc = ssrc;
        this.scalabilityMode = scalabilityMode;
        this.mappedSsrc = mappedSsrc;
    }
    pack(builder) {
        const rid = (this.rid !== null ? builder.createString(this.rid) : 0);
        const scalabilityMode = (this.scalabilityMode !== null ? builder.createString(this.scalabilityMode) : 0);
        return EncodingMapping.createEncodingMapping(builder, rid, this.ssrc, scalabilityMode, this.mappedSsrc);
    }
}
exports.EncodingMappingT = EncodingMappingT;
