// mediasoup/node/lib/fbs/rtp-parameters/rtcp-feedback.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcpFeedbackT = exports.RtcpFeedback = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
class RtcpFeedback {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsRtcpFeedback(bb, obj) {
        return (obj || new RtcpFeedback()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsRtcpFeedback(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new RtcpFeedback()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    type(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    parameter(optionalEncoding) {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.__string(this.bb_pos + offset, optionalEncoding) : null;
    }
    static startRtcpFeedback(builder) {
        builder.startObject(2);
    }
    static addType(builder, typeOffset) {
        builder.addFieldOffset(0, typeOffset, 0);
    }
    static addParameter(builder, parameterOffset) {
        builder.addFieldOffset(1, parameterOffset, 0);
    }
    static endRtcpFeedback(builder) {
        const offset = builder.endObject();
        builder.requiredField(offset, 4); // type
        return offset;
    }
    static createRtcpFeedback(builder, typeOffset, parameterOffset) {
        RtcpFeedback.startRtcpFeedback(builder);
        RtcpFeedback.addType(builder, typeOffset);
        RtcpFeedback.addParameter(builder, parameterOffset);
        return RtcpFeedback.endRtcpFeedback(builder);
    }
    unpack() {
        return new RtcpFeedbackT(this.type(), this.parameter());
    }
    unpackTo(_o) {
        _o.type = this.type();
        _o.parameter = this.parameter();
    }
}
exports.RtcpFeedback = RtcpFeedback;
class RtcpFeedbackT {
    type;
    parameter;
    constructor(type = null, parameter = null) {
        this.type = type;
        this.parameter = parameter;
    }
    pack(builder) {
        const type = (this.type !== null ? builder.createString(this.type) : 0);
        const parameter = (this.parameter !== null ? builder.createString(this.parameter) : 0);
        return RtcpFeedback.createRtcpFeedback(builder, type, parameter);
    }
}
exports.RtcpFeedbackT = RtcpFeedbackT;
