// mediasoup/node/lib/test/test-werift-sctp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const node_dgram_1 = require("node:dgram");
const werift_sctp_1 = require("werift-sctp");
const mediasoup = __importStar(require("../"));
const enhancedEvents_1 = require("../enhancedEvents");
const ctx = {};
beforeEach(async () => {
    ctx.worker = await mediasoup.createWorker();
    ctx.router = await ctx.worker.createRouter();
    ctx.plainTransport = await ctx.router.createPlainTransport({
        // https://github.com/nodejs/node/issues/14900.
        listenIp: '127.0.0.1',
        // So we don't need to call plainTransport.connect().
        comedia: true,
        enableSctp: true,
    });
    // Create an explicit SCTP outgoing stream id.
    ctx.sctpSendStreamId = 123;
    ctx.sctpClient = werift_sctp_1.SCTP.client((0, werift_sctp_1.createUdpTransport)((0, node_dgram_1.createSocket)('udp4'), {
        port: ctx.plainTransport.tuple.localPort,
        address: ctx.plainTransport.tuple.localAddress,
    }));
    // Create a DataProducer with the corresponding SCTP stream id.
    ctx.dataProducer = await ctx.plainTransport.produceData({
        sctpStreamParameters: {
            streamId: ctx.sctpSendStreamId,
            ordered: true,
        },
        label: 'node-sctp',
        protocol: 'foo & bar 😀😀😀',
    });
    // Create a DataConsumer to receive messages from the DataProducer over the
    // same plainTransport.
    ctx.dataConsumer = await ctx.plainTransport.consumeData({
        dataProducerId: ctx.dataProducer.id,
    });
    let connectionTimeoutTimer;
    await Promise.race([
        // Wait for SCTP to become connected in both the PlainTransport and in the
        // werift-sctp client.
        Promise.all([
            // Connect werift-sctp client (this resolves once SCTP is connected).
            ctx.sctpClient.start(5000),
            // This resolves once connected too.
            ctx.sctpClient.stateChanged.connected.asPromise(),
            // Wait for SCTP state in the mediasoup PlainTransport to be "connected".
            new Promise((resolve, reject) => {
                if (ctx.plainTransport?.sctpState === 'connected') {
                    resolve();
                }
                else {
                    ctx.plainTransport?.on('sctpstatechange', state => {
                        if (state === 'connected') {
                            resolve();
                        }
                        else if (state === 'failed' || state === 'closed') {
                            reject(new Error('SCTP connection in PlainTransport failed or was closed'));
                        }
                    });
                }
            }),
        ]),
        new Promise((resolve, reject) => {
            connectionTimeoutTimer = setTimeout(() => reject(new Error('SCTP connection timeout')), 3000);
        }),
    ]);
    clearTimeout(connectionTimeoutTimer);
}, 5000);
afterEach(async () => {
    await ctx.sctpClient?.stop();
    ctx.sctpClient?.transport.close();
    ctx.worker?.close();
    if (ctx.worker?.subprocessClosed === false) {
        await (0, enhancedEvents_1.enhancedOnce)(ctx.worker, 'subprocessclose');
    }
});
test('SCTP state is connected', async () => {
    expect(ctx.plainTransport.sctpState).toBe('connected');
    expect(ctx.plainTransport.sctpNegotiatedCapabilities).toEqual({
        negotiatedMaxOutboundStreams: 65535,
        negotiatedMaxInboundStreams: 65535,
    });
    expect(ctx.sctpClient.associationState).toBe(werift_sctp_1.SCTP_STATE.ESTABLISHED);
    const dump = await ctx.plainTransport.dump();
    expect(dump.sctpState).toBe('connected');
    expect(dump.sctpNegotiatedCapabilities).toEqual({
        negotiatedMaxOutboundStreams: 65535,
        negotiatedMaxInboundStreams: 65535,
    });
});
test('ordered DataProducer delivers all SCTP messages to the DataConsumer', async () => {
    const numMessages = 200;
    let sentMessageBytes = 0;
    let recvMessageBytes = 0;
    let numSentMessages = 0;
    let numReceivedMessages = 0;
    // It must be zero because it's the first DataConsumer on the plainTransport.
    expect(ctx.dataConsumer.sctpStreamParameters?.streamId).toBe(0);
    await new Promise((resolve, reject) => {
        sendNextMessage();
        function sendNextMessage() {
            const id = ++numSentMessages;
            const data = Buffer.from(String(id));
            let ppid;
            // Set ppid of type WebRTC DataChannel string.
            if (id < numMessages / 2) {
                ppid = werift_sctp_1.WEBRTC_PPID.STRING;
            }
            // Set ppid of type WebRTC DataChannel binary.
            else {
                ppid = werift_sctp_1.WEBRTC_PPID.BINARY;
            }
            void ctx.sctpClient.send(ctx.sctpSendStreamId, ppid, data);
            sentMessageBytes += data.byteLength;
            if (id < numMessages) {
                sendNextMessage();
            }
        }
        ctx.sctpClient.onReceive.subscribe((streamId, ppid, data) => {
            // `streamId`  must be zero because it's the first SCTP incoming stream
            // (so first DataConsumer).
            if (streamId !== 0) {
                reject(new Error(`streamId should be 0 but it is ${streamId}`));
                return;
            }
            ++numReceivedMessages;
            recvMessageBytes += data.byteLength;
            const id = Number(data.toString('utf8'));
            if (id !== numReceivedMessages) {
                reject(new Error(`id ${id} in message should match numReceivedMessages ${numReceivedMessages}`));
            }
            else if (id === numMessages) {
                resolve();
            }
            else if (id < numMessages / 2 && ppid !== werift_sctp_1.WEBRTC_PPID.STRING) {
                reject(new Error(`ppid in message with id ${id} should be ${werift_sctp_1.WEBRTC_PPID.STRING} but it is ${ppid}`));
            }
            else if (id > numMessages / 2 && ppid !== werift_sctp_1.WEBRTC_PPID.BINARY) {
                reject(new Error(`ppid in message with id ${id} should be ${werift_sctp_1.WEBRTC_PPID.BINARY} but it is ${ppid}`));
                return;
            }
        });
    });
    expect(numSentMessages).toBe(numMessages);
    expect(numReceivedMessages).toBe(numMessages);
    expect(recvMessageBytes).toBe(sentMessageBytes);
    await expect(ctx.dataProducer.getStats()).resolves.toMatchObject([
        {
            type: 'data-producer',
            label: ctx.dataProducer.label,
            protocol: ctx.dataProducer.protocol,
            messagesReceived: numMessages,
            bytesReceived: sentMessageBytes,
        },
    ]);
    await expect(ctx.dataConsumer.getStats()).resolves.toMatchObject([
        {
            type: 'data-consumer',
            label: ctx.dataConsumer.label,
            protocol: ctx.dataConsumer.protocol,
            messagesSent: numMessages,
            bytesSent: recvMessageBytes,
        },
    ]);
    expect(await ctx.dataConsumer.getBufferedAmount()).toBe(0);
}, 10000);
test('can send messages from the SCTP DataConsumer', async () => {
    expect(await ctx.dataConsumer.getBufferedAmount()).toBe(0);
    // We can send messages directly from the DataConsumer and it should return
    // current buffered amount, so 0 bytes.
    await expect(ctx.dataConsumer.send('foo')).resolves.toBe(0);
}, 2000);
