// mediasoup/node/lib/test/test-ActiveSpeakerObserver.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const mediasoup = __importStar(require("../"));
const enhancedEvents_1 = require("../enhancedEvents");
const errors_1 = require("../errors");
const utils = __importStar(require("../utils"));
const ctx = {
    mediaCodecs: utils.deepFreeze([
        {
            kind: 'audio',
            mimeType: 'audio/opus',
            clockRate: 48000,
            channels: 2,
            parameters: {
                useinbandfec: 1,
                foo: 'bar',
            },
        },
    ]),
};
beforeEach(async () => {
    ctx.worker = await mediasoup.createWorker();
    ctx.router = await ctx.worker.createRouter({ mediaCodecs: ctx.mediaCodecs });
});
afterEach(async () => {
    ctx.worker?.close();
    if (ctx.worker?.subprocessClosed === false) {
        await (0, enhancedEvents_1.enhancedOnce)(ctx.worker, 'subprocessclose');
    }
});
test('router.createActiveSpeakerObserver() succeeds', async () => {
    const onObserverNewRtpObserver = jest.fn();
    ctx.router.observer.once('newrtpobserver', onObserverNewRtpObserver);
    const activeSpeakerObserver = await ctx.router.createActiveSpeakerObserver();
    expect(onObserverNewRtpObserver).toHaveBeenCalledTimes(1);
    expect(onObserverNewRtpObserver).toHaveBeenCalledWith(activeSpeakerObserver);
    expect(typeof activeSpeakerObserver.id).toBe('string');
    expect(activeSpeakerObserver.closed).toBe(false);
    expect(activeSpeakerObserver.type).toBe('activespeaker');
    expect(activeSpeakerObserver.paused).toBe(false);
    expect(activeSpeakerObserver.appData).toEqual({});
    await expect(ctx.router.dump()).resolves.toMatchObject({
        rtpObserverIds: [activeSpeakerObserver.id],
    });
}, 2000);
test('router.createActiveSpeakerObserver() with wrong arguments rejects with TypeError', async () => {
    await expect(ctx.router.createActiveSpeakerObserver(
    // @ts-expect-error --- Testing purposes.
    { interval: false })).rejects.toThrow(TypeError);
    await expect(ctx.router.createActiveSpeakerObserver(
    // @ts-expect-error --- Testing purposes.
    { appData: 'NOT-AN-OBJECT' })).rejects.toThrow(TypeError);
}, 2000);
test('activeSpeakerObserver.pause() and resume() succeed', async () => {
    const activeSpeakerObserver = await ctx.router.createActiveSpeakerObserver();
    await activeSpeakerObserver.pause();
    expect(activeSpeakerObserver.paused).toBe(true);
    await activeSpeakerObserver.resume();
    expect(activeSpeakerObserver.paused).toBe(false);
}, 2000);
test('activeSpeakerObserver.close() succeeds', async () => {
    const activeSpeakerObserver = await ctx.router.createActiveSpeakerObserver({
        interval: 500,
    });
    let dump = await ctx.router.dump();
    expect(dump.rtpObserverIds.length).toBe(1);
    activeSpeakerObserver.close();
    expect(activeSpeakerObserver.closed).toBe(true);
    dump = await ctx.router.dump();
    expect(dump.rtpObserverIds.length).toBe(0);
    await expect(activeSpeakerObserver.pause()).rejects.toThrow(errors_1.NotFoundError);
}, 2000);
test('ActiveSpeakerObserver emits "routerclose" if Router is closed', async () => {
    const activeSpeakerObserver = await ctx.router.createActiveSpeakerObserver();
    const promise = (0, enhancedEvents_1.enhancedOnce)(activeSpeakerObserver, 'routerclose');
    ctx.router.close();
    await promise;
    expect(activeSpeakerObserver.closed).toBe(true);
    await expect(activeSpeakerObserver.pause()).rejects.toThrow(errors_1.NotFoundError);
}, 2000);
test('ActiveSpeakerObserver emits "routerclose" if Worker is closed', async () => {
    const activeSpeakerObserver = await ctx.router.createActiveSpeakerObserver();
    const promise = (0, enhancedEvents_1.enhancedOnce)(activeSpeakerObserver, 'routerclose');
    ctx.worker.close();
    await promise;
    expect(activeSpeakerObserver.closed).toBe(true);
    await expect(activeSpeakerObserver.pause()).rejects.toThrow(errors_1.WorkerClosedError);
}, 2000);
