// mediasoup/node/lib/rtpStreamStatsFbsUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseRtpStreamStats = parseRtpStreamStats;
exports.parseRtpStreamRecvStats = parseRtpStreamRecvStats;
exports.parseSendStreamStats = parseSendStreamStats;
const FbsRtpStream = __importStar(require("./fbs/rtp-stream"));
const FbsRtpParameters = __importStar(require("./fbs/rtp-parameters"));
function parseRtpStreamStats(binary) {
    if (binary.dataType() === FbsRtpStream.StatsData.RecvStats) {
        return parseRtpStreamRecvStats(binary);
    }
    else {
        return parseSendStreamStats(binary);
    }
}
function parseRtpStreamRecvStats(binary) {
    const recvStats = new FbsRtpStream.RecvStats();
    const baseStats = new FbsRtpStream.BaseStats();
    binary.data(recvStats);
    recvStats.base().data(baseStats);
    const base = parseBaseStreamStats(baseStats);
    return {
        ...base,
        type: 'inbound-rtp',
        byteCount: Number(recvStats.byteCount()),
        packetCount: Number(recvStats.packetCount()),
        bitrate: Number(recvStats.bitrate()),
        bitrateByLayer: parseBitrateByLayer(recvStats),
    };
}
function parseSendStreamStats(binary) {
    const sendStats = new FbsRtpStream.SendStats();
    const baseStats = new FbsRtpStream.BaseStats();
    binary.data(sendStats);
    sendStats.base().data(baseStats);
    const base = parseBaseStreamStats(baseStats);
    return {
        ...base,
        type: 'outbound-rtp',
        byteCount: Number(sendStats.byteCount()),
        packetCount: Number(sendStats.packetCount()),
        bitrate: Number(sendStats.bitrate()),
    };
}
function parseBaseStreamStats(binary) {
    return {
        timestamp: Number(binary.timestamp()),
        ssrc: binary.ssrc(),
        rtxSsrc: binary.rtxSsrc() ?? undefined,
        rid: binary.rid() || undefined,
        kind: binary.kind() === FbsRtpParameters.MediaKind.AUDIO ? 'audio' : 'video',
        mimeType: binary.mimeType(),
        packetsLost: Number(binary.packetsLost()),
        fractionLost: Number(binary.fractionLost()),
        jitter: Number(binary.jitter()),
        packetsDiscarded: Number(binary.packetsDiscarded()),
        packetsRetransmitted: Number(binary.packetsRetransmitted()),
        packetsRepaired: Number(binary.packetsRepaired()),
        nackCount: Number(binary.nackCount()),
        nackPacketCount: Number(binary.nackPacketCount()),
        pliCount: Number(binary.pliCount()),
        firCount: Number(binary.firCount()),
        roundTripTime: binary.roundTripTime(),
        rtxPacketsDiscarded: binary.rtxPacketsDiscarded()
            ? Number(binary.rtxPacketsDiscarded())
            : undefined,
        score: binary.score(),
    };
}
function parseBitrateByLayer(binary) {
    if (binary.bitrateByLayerLength() === 0) {
        return {};
    }
    const bitRateByLayer = {};
    for (let i = 0; i < binary.bitrateByLayerLength(); ++i) {
        const layer = binary.bitrateByLayer(i).layer();
        const bitrate = binary.bitrateByLayer(i).bitrate();
        bitRateByLayer[layer] = Number(bitrate);
    }
    return bitRateByLayer;
}
