// mediasoup/node/lib/srtpParametersFbsUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.cryptoSuiteFromFbs = cryptoSuiteFromFbs;
exports.cryptoSuiteToFbs = cryptoSuiteToFbs;
exports.parseSrtpParameters = parseSrtpParameters;
exports.serializeSrtpParameters = serializeSrtpParameters;
const FbsSrtpParameters = __importStar(require("./fbs/srtp-parameters"));
function cryptoSuiteFromFbs(binary) {
    switch (binary) {
        case FbsSrtpParameters.SrtpCryptoSuite.AEAD_AES_256_GCM: {
            return 'AEAD_AES_256_GCM';
        }
        case FbsSrtpParameters.SrtpCryptoSuite.AEAD_AES_128_GCM: {
            return 'AEAD_AES_128_GCM';
        }
        case FbsSrtpParameters.SrtpCryptoSuite.AES_CM_128_HMAC_SHA1_80: {
            return 'AES_CM_128_HMAC_SHA1_80';
        }
        case FbsSrtpParameters.SrtpCryptoSuite.AES_CM_128_HMAC_SHA1_32: {
            return 'AES_CM_128_HMAC_SHA1_32';
        }
    }
}
function cryptoSuiteToFbs(cryptoSuite) {
    switch (cryptoSuite) {
        case 'AEAD_AES_256_GCM': {
            return FbsSrtpParameters.SrtpCryptoSuite.AEAD_AES_256_GCM;
        }
        case 'AEAD_AES_128_GCM': {
            return FbsSrtpParameters.SrtpCryptoSuite.AEAD_AES_128_GCM;
        }
        case 'AES_CM_128_HMAC_SHA1_80': {
            return FbsSrtpParameters.SrtpCryptoSuite.AES_CM_128_HMAC_SHA1_80;
        }
        case 'AES_CM_128_HMAC_SHA1_32': {
            return FbsSrtpParameters.SrtpCryptoSuite.AES_CM_128_HMAC_SHA1_32;
        }
        default: {
            throw new TypeError(`invalid SrtpCryptoSuite: ${cryptoSuite}`);
        }
    }
}
function parseSrtpParameters(binary) {
    return {
        cryptoSuite: cryptoSuiteFromFbs(binary.cryptoSuite()),
        keyBase64: binary.keyBase64(),
    };
}
function serializeSrtpParameters(builder, srtpParameters) {
    const keyBase64Offset = builder.createString(srtpParameters.keyBase64);
    return FbsSrtpParameters.SrtpParameters.createSrtpParameters(builder, cryptoSuiteToFbs(srtpParameters.cryptoSuite), keyBase64Offset);
}
