// mediasoup/node/lib/fbs/rtp-stream/recv-stats.js
"use strict";
// automatically generated by the FlatBuffers compiler, do not modify
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RecvStatsT = exports.RecvStats = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/no-non-null-assertion */
const flatbuffers = __importStar(require("flatbuffers"));
const bitrate_by_layer_1 = require("../../fbs/rtp-stream/bitrate-by-layer");
const stats_1 = require("../../fbs/rtp-stream/stats");
class RecvStats {
    bb = null;
    bb_pos = 0;
    __init(i, bb) {
        this.bb_pos = i;
        this.bb = bb;
        return this;
    }
    static getRootAsRecvStats(bb, obj) {
        return (obj || new RecvStats()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    static getSizePrefixedRootAsRecvStats(bb, obj) {
        bb.setPosition(bb.position() + flatbuffers.SIZE_PREFIX_LENGTH);
        return (obj || new RecvStats()).__init(bb.readInt32(bb.position()) + bb.position(), bb);
    }
    base(obj) {
        const offset = this.bb.__offset(this.bb_pos, 4);
        return offset ? (obj || new stats_1.Stats()).__init(this.bb.__indirect(this.bb_pos + offset), this.bb) : null;
    }
    packetCount() {
        const offset = this.bb.__offset(this.bb_pos, 6);
        return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt('0');
    }
    byteCount() {
        const offset = this.bb.__offset(this.bb_pos, 8);
        return offset ? this.bb.readUint64(this.bb_pos + offset) : BigInt('0');
    }
    bitrate() {
        const offset = this.bb.__offset(this.bb_pos, 10);
        return offset ? this.bb.readUint32(this.bb_pos + offset) : 0;
    }
    bitrateByLayer(index, obj) {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? (obj || new bitrate_by_layer_1.BitrateByLayer()).__init(this.bb.__indirect(this.bb.__vector(this.bb_pos + offset) + index * 4), this.bb) : null;
    }
    bitrateByLayerLength() {
        const offset = this.bb.__offset(this.bb_pos, 12);
        return offset ? this.bb.__vector_len(this.bb_pos + offset) : 0;
    }
    static startRecvStats(builder) {
        builder.startObject(5);
    }
    static addBase(builder, baseOffset) {
        builder.addFieldOffset(0, baseOffset, 0);
    }
    static addPacketCount(builder, packetCount) {
        builder.addFieldInt64(1, packetCount, BigInt('0'));
    }
    static addByteCount(builder, byteCount) {
        builder.addFieldInt64(2, byteCount, BigInt('0'));
    }
    static addBitrate(builder, bitrate) {
        builder.addFieldInt32(3, bitrate, 0);
    }
    static addBitrateByLayer(builder, bitrateByLayerOffset) {
        builder.addFieldOffset(4, bitrateByLayerOffset, 0);
    }
    static createBitrateByLayerVector(builder, data) {
        builder.startVector(4, data.length, 4);
        for (let i = data.length - 1; i >= 0; i--) {
            builder.addOffset(data[i]);
        }
        return builder.endVector();
    }
    static startBitrateByLayerVector(builder, numElems) {
        builder.startVector(4, numElems, 4);
    }
    static endRecvStats(builder) {
        const offset = builder.endObject();
        builder.requiredField(offset, 4); // base
        builder.requiredField(offset, 12); // bitrate_by_layer
        return offset;
    }
    static createRecvStats(builder, baseOffset, packetCount, byteCount, bitrate, bitrateByLayerOffset) {
        RecvStats.startRecvStats(builder);
        RecvStats.addBase(builder, baseOffset);
        RecvStats.addPacketCount(builder, packetCount);
        RecvStats.addByteCount(builder, byteCount);
        RecvStats.addBitrate(builder, bitrate);
        RecvStats.addBitrateByLayer(builder, bitrateByLayerOffset);
        return RecvStats.endRecvStats(builder);
    }
    unpack() {
        return new RecvStatsT((this.base() !== null ? this.base().unpack() : null), this.packetCount(), this.byteCount(), this.bitrate(), this.bb.createObjList(this.bitrateByLayer.bind(this), this.bitrateByLayerLength()));
    }
    unpackTo(_o) {
        _o.base = (this.base() !== null ? this.base().unpack() : null);
        _o.packetCount = this.packetCount();
        _o.byteCount = this.byteCount();
        _o.bitrate = this.bitrate();
        _o.bitrateByLayer = this.bb.createObjList(this.bitrateByLayer.bind(this), this.bitrateByLayerLength());
    }
}
exports.RecvStats = RecvStats;
class RecvStatsT {
    base;
    packetCount;
    byteCount;
    bitrate;
    bitrateByLayer;
    constructor(base = null, packetCount = BigInt('0'), byteCount = BigInt('0'), bitrate = 0, bitrateByLayer = []) {
        this.base = base;
        this.packetCount = packetCount;
        this.byteCount = byteCount;
        this.bitrate = bitrate;
        this.bitrateByLayer = bitrateByLayer;
    }
    pack(builder) {
        const base = (this.base !== null ? this.base.pack(builder) : 0);
        const bitrateByLayer = RecvStats.createBitrateByLayerVector(builder, builder.createObjectOffsetList(this.bitrateByLayer));
        return RecvStats.createRecvStats(builder, base, this.packetCount, this.byteCount, this.bitrate, bitrateByLayer);
    }
}
exports.RecvStatsT = RecvStatsT;
